src/mpi/comm/comm_impl.c

```diff
@@ -726,25 +726,42 @@ static int get_tag_from_stringtag(const char *stringtag)
     return hash % (MPIR_Process.attrs.tag_ub);
 }
 
+static bool is_world_group(MPIR_Group * group_ptr)
+{
+    return (group_ptr->size == MPIR_Process.size && group_ptr->size > 1);
+}
+
+static bool is_self_group(MPIR_Group * group_ptr)
+{
+    return (group_ptr->size == 1);
+}
+
 int MPIR_Comm_create_from_group_impl(MPIR_Group * group_ptr, const char *stringtag,
                                      MPIR_Info * info_ptr, MPIR_Errhandler * errhan_ptr,
                                      MPIR_Comm ** p_newcom_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
+    int use_comm_world = 0;
 
-    /* This implementation assumes an internal MPI_COMM_WORLD already exist.
-     *
-     * Refer to TODOs below inside the branches:
-     * The next implementation will relax but assume the first call to this function will
-     * either from a "world" pset or "self" pset, during former a comm world will be created.
-     *
-     * The final implementation will further relax and allow partial initializations.
-     */
-
-    /* NOTE: tag will be used with MPIR_TAG_COLL_BIT on, ref. MPIR_Get_contextid_sparse_group */
-    int tag = get_tag_from_stringtag(stringtag);
+    /* NOTE: only world or self is supported without first establishing comm_world.  */
 
+    MPL_initlock_lock(&MPIR_init_lock);
     if (MPIR_Process.comm_world) {
+        use_comm_world = 1;
+    } else if (is_world_group(group_ptr)) {
+        mpi_errno = MPIR_init_comm_world();
+        MPIR_ERR_CHECK(mpi_errno);
+        use_comm_world = 1;
+    } else if (!MPIR_Process.comm_self && is_self_group(group_ptr)) {
+        mpi_errno = MPIR_init_comm_self();
+        MPIR_ERR_CHECK(mpi_errno);
+    }
```

Comment:
Is it necessary to have these special cases for world/self? Aren't they all just communicators created from the pool of processes that we know about from the AV? What still requires creating these specific communicators?

Reply:
First, in the world model, these built-in communicators need to be created internally thus we will need to store the pointers internally anyway and manage to release them upon finalizing.

Second, since we have the built-in comms anyway, having user comms all derived from built-in-comms simplifies the design. We will only need to worry about bootstrapping built-in comms.

The first world-comm needs to bootstrap the whole world connections. but we should delay it and not to bootstrap the world when self-comm is created, thus the above quoted code.

Reply:
I guess this function could be used for the sessions model or the world model so maybe this is still fine. I was just checking if we're creating these communicators when we don't need to when using the sessions model.

Reply:
Even if we restrict this function to be used in the sessions model, the world model still can exist simultaneously, so the situation doesn't change. Design-wise, while we could make this function built-in comm mutual, it will be more complicated. We will need to track world initialization states anyway, so replacing `MPIR_Process.comm_world` with another state variable does not gain anything. We'll need a separate context_id allocation algorithm that works without an existing communicator. And we'll have to modify and alter the current world model code to account for the session model. So I will delay the complications until necessary.